File: apps/api/app/schemas/memory.py
"""Memory request/response schemas."""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field


# ── Memory ────────────────────────────────────────────────────────────────


class MemoryCreateRequest(BaseModel):
    """Schema for creating a memory."""

    title: str = Field(..., min_length=1, max_length=500)
    content: str = Field(..., min_length=1)
    summary: str | None = Field(None, max_length=2000)
    memory_type: str = Field(..., description="Memory type enum value")
    project_id: uuid.UUID | None = None
    source: str | None = Field(None, max_length=255)
    source_url: str | None = None
    tags: list[str] = Field(default_factory=list)
    importance_score: float = Field(default=0.5, ge=0.0, le=1.0)
    confidence: float = Field(default=1.0, ge=0.0, le=1.0)
    metadata_json: dict | None = None
    chunking_strategy: str = Field(
        default="recursive",
        description="Chunking strategy: sentence, paragraph, recursive",
    )


class MemoryUpdateRequest(BaseModel):
    """Schema for updating a memory."""

    title: str | None = Field(None, min_length=1, max_length=500)
    content: str | None = Field(None, min_length=1)
    summary: str | None = Field(None, max_length=2000)
    importance_score: float | None = Field(None, ge=0.0, le=1.0)
    confidence: float | None = Field(None, ge=0.0, le=1.0)
    status: str | None = None
    tags: list[str] | None = None
    metadata_json: dict | None = None


class MemoryResponse(BaseModel):
    """Schema for memory data in API responses."""

    id: uuid.UUID
    owner_id: uuid.UUID
    workspace_id: uuid.UUID
    project_id: uuid.UUID | None
    title: str
    content: str
    summary: str | None
    embedding_id: str | None
    memory_type: str
    source: str | None
    source_url: str | None
    metadata_json: dict | None
    importance_score: float
    confidence: float
    status: str
    access_count: int
    created_at: datetime
    updated_at: datetime
    accessed_at: datetime | None
    tags: list[str] = []

    model_config = {"from_attributes": True}


class MemoryListResponse(BaseModel):
    """Paginated memory list response."""

    memories: list[MemoryResponse]
    total: int
    page: int
    page_size: int


# ── Search ────────────────────────────────────────────────────────────────


class MemorySearchRequest(BaseModel):
    """Schema for semantic memory search."""

    query: str = Field(..., min_length=1, max_length=1000)
    workspace_id: uuid.UUID
    project_id: uuid.UUID | None = None
    memory_type: str | None = None
    tags: list[str] = Field(default_factory=list)
    top_k: int = Field(default=10, ge=1, le=100)
    min_importance: float | None = Field(None, ge=0.0, le=1.0)


class MemorySearchResult(BaseModel):
    """A single search result with similarity score."""

    memory: MemoryResponse
    score: float
    chunk_content: str | None = None


class MemorySearchResponse(BaseModel):
    """Search results response."""

    results: list[MemorySearchResult]
    query: str
    total: int


# ── Similarity ────────────────────────────────────────────────────────────


class SimilarityRequest(BaseModel):
    """Schema for finding similar memories."""

    memory_id: uuid.UUID
    top_k: int = Field(default=5, ge=1, le=50)


class SimilarityResponse(BaseModel):
    """Similarity search results."""

    source_memory: MemoryResponse
    similar_memories: list[MemorySearchResult]


# ── Graph ─────────────────────────────────────────────────────────────────


class GraphNodeResponse(BaseModel):
    """A node in the memory graph."""

    id: str
    label: str
    type: str
    memory_type: str | None = None
    metadata: dict | None = None


class GraphEdgeResponse(BaseModel):
    """An edge in the memory graph."""

    id: str
    source: str
    target: str
    label: str
    weight: float = 1.0


class GraphResponse(BaseModel):
    """Full graph response for visualization."""

    nodes: list[GraphNodeResponse]
    edges: list[GraphEdgeResponse]

File: apps/web/src/types/memory.ts
/* ── Memory Types ──────────────────────────────────────────────────── */

export interface Memory {
  id: string;
  owner_id: string;
  workspace_id: string;
  project_id: string | null;
  title: string;
  content: string;
  summary: string | null;
  embedding_id: string | null;
  memory_type: string;
  source: string | null;
  source_url: string | null;
  metadata_json: Record<string, unknown> | null;
  importance_score: number;
  confidence: number;
  status: string;
  access_count: number;
  created_at: string;
  updated_at: string;
  accessed_at: string | null;
  tags: string[];
}

export interface MemorySearchResult {
  memory: Memory;
  score: number;
  chunk_content: string | null;
}

export interface MemoryListResponse {
  memories: Memory[];
  total: number;
  page: number;
  page_size: number;
}

export interface MemorySearchResponse {
  results: MemorySearchResult[];
  query: string;
  total: number;
}

export interface GraphNode {
  id: string;
  label: string;
  type: string;
  memory_type?: string;
  metadata?: Record<string, unknown>;
}

export interface GraphEdge {
  id: string;
  source: string;
  target: string;
  label: string;
  weight?: number;
}

export interface GraphData {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface MemoryCreateRequest {
  title: string;
  content: string;
  summary?: string;
  memory_type?: string;
  project_id?: string;
  source?: string;
  tags?: string[];
  importance_score?: number;
  chunking_strategy?: string;
}

export interface MemoryUpdateRequest {
  title?: string;
  content?: string;
  summary?: string;
  memory_type?: string;
  source?: string;
  tags?: string[];
  importance_score?: number;
  status?: string;
}

/* ── Memory Type Options ──────────────────────────────────────────── */

export const MEMORY_TYPES = [
  { value: "conversation", label: "Conversation" },
  { value: "knowledge", label: "Knowledge" },
  { value: "document", label: "Document" },
  { value: "task", label: "Task" },
  { value: "project", label: "Project" },
  { value: "preference", label: "Preference" },
  { value: "event", label: "Event" },
  { value: "observation", label: "Observation" },
] as const;

export const MEMORY_TYPE_COLORS: Record<string, string> = {
  conversation: "#3b82f6",
  knowledge: "#8b5cf6",
  document: "#10b981",
  task: "#f59e0b",
  project: "#6366f1",
  preference: "#ec4899",
  event: "#ef4444",
  observation: "#06b6d4",
};
